src/libraries/System.Text.Json/src/System/Text/Json/Document/JsonDocument.cs

```diff
@@ -1096,6 +1096,86 @@ private static void Parse(
             database.CompleteAllocations();
         }
 
+        private static void ValidateDocument(JsonDocument document, bool allowDuplicateProperties)
+        {
+            if (!allowDuplicateProperties &&
+                document.RootElement.ValueKind is JsonValueKind.Array or JsonValueKind.Object)
+            {
+                ValidateDuplicateProperties(document);
+            }
+        }
+
+        private static void ValidateDuplicateProperties(JsonDocument document)
+        {
+            Debug.Assert(document.RootElement.ValueKind is JsonValueKind.Array or JsonValueKind.Object);
+
+            using PropertyNameSet propertyNameSet = new PropertyNameSet();
+
+            Stack<int> traversalPath = new Stack<int>();
+            int? databaseIndexOflastProcessedChild = null;
+
+            traversalPath.Push(document.RootElement.MetadataDbIndex);
+
+            do
+            {
+                JsonElement curr = new JsonElement(document, traversalPath.Peek());
```

Comment:
Why can't we just traverse the `JsonElement` with the usual publicly available APIs?